compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala

```diff
@@ -1209,20 +1209,29 @@ object messages {
            |""".stripMargin
   }
 
-  case class OverloadedOrRecursiveMethodNeedsResultType(tree: Names.TermName)(implicit ctx: Context)
+  case class OverloadedOrRecursiveMethodNeedsResultType private (termName: String)(implicit ctx: Context)
   extends Message(OverloadedOrRecursiveMethodNeedsResultTypeID) {
     val kind = "Syntax"
-    val msg = hl"""overloaded or recursive method ${tree} needs return type"""
+    val msg = hl"""overloaded or recursive method $termName needs return type"""
     val explanation =
-      hl"""Case 1: ${tree} is overloaded
-          |If there are multiple methods named `${tree.name}` and at least one definition of
+      hl"""Case 1: $termName is overloaded
+          |If there are multiple methods named `$termName` and at least one definition of
           |it calls another, you need to specify the calling method's return type.
           |
-          |Case 2: ${tree} is recursive
-          |If `${tree.name}` calls itself on any path, you need to specify its return type.
+          |Case 2: $termName is recursive
+          |If `$termName` calls itself on any path, you need to specify its return type.
           |""".stripMargin
   }
 
+  object OverloadedOrRecursiveMethodNeedsResultType {
+    @specialized def apply[T >: Trees.Untyped](tree: NameTree[T])(implicit ctx: Context)
+    : OverloadedOrRecursiveMethodNeedsResultType =
+      OverloadedOrRecursiveMethodNeedsResultType(tree.name.toString)(ctx)
+    def apply(symbol: Symbol)(implicit ctx: Context)
+    : OverloadedOrRecursiveMethodNeedsResultType =
+      OverloadedOrRecursiveMethodNeedsResultType(symbol.name.toString)(ctx)
```

Comment:
There's a "Showable" type class that can be accessed for both trees and symbols via `.show`